tests/cpp/tree/gpu_hist/test_histogram.cu

```diff
@@ -95,11 +94,25 @@ TEST(Histogram, GPUDeterministic) {
   std::vector<int> shm_sizes{48 * 1024, 64 * 1024, 160 * 1024};
   for (bool is_dense : is_dense_array) {
     for (int shm_size : shm_sizes) {
-      TestDeterministicHistogram<GradientPairPrecise>(is_dense, shm_size);
+      TestDeterministicHistogram(is_dense, shm_size);
     }
   }
 }
 
+void ValidateCategoricalHistogram(size_t n_categories, common::Span<GradientPairInt64> onehot,
```

Comment:
Can we merge the CPU and GPU versions back if we can deduplicate the gradient definition or use a template for this function?

Reply:
As mentioned above, it's intentional to force the integer gradients to be handled differently.